From d6bb93ee1a5ecae8df55b82747c7375b6d12c1c6 Mon Sep 17 00:00:00 2001
From: Mateusz Chudyk <mateusz.chudyk@intel.com>
Date: Fri, 24 Feb 2023 10:13:57 +0000
Subject: [PATCH]  PromoteBools - use replaceAllUsesWith when promoted type is
 equal to the origin one

PromoteBools - use replaceAllUsesWith when promoted type is equal to the origin one
---
 IGC/AdaptorOCL/preprocess_spvir/PromoteBools.cpp | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/IGC/AdaptorOCL/preprocess_spvir/PromoteBools.cpp b/IGC/AdaptorOCL/preprocess_spvir/PromoteBools.cpp
index b5bf0af2a9f..a28ab22c77a 100644
--- a/IGC/AdaptorOCL/preprocess_spvir/PromoteBools.cpp
+++ b/IGC/AdaptorOCL/preprocess_spvir/PromoteBools.cpp
@@ -442,11 +442,18 @@ Value* PromoteBools::getOrCreatePromotedValue(Value* value)
     if (newValue != value)
     {
         promotedValuesCache[value] = newValue;
-        for (const auto& user : value->users())
+        if (value->getType() == newValue->getType())
         {
-            if (!wasPromoted(user))
+            value->replaceAllUsesWith(newValue);
+        }
+        else
+        {
+            for (const auto& user : value->users())
             {
-                promotionQueue.push(user);
+                if (!wasPromoted(user))
+                {
+                    promotionQueue.push(user);
+                }
             }
         }
     }
